### ISSUE TITLE
Improve hax::main macro

### ISSUE THREAD

**Issue by 7ap** [hax-rs/hax#2]
- Manual mapping? (internal)
- Cleaner hooking? (internal)
  - `#[hax::main(init=setup_hooks)]`
- Try to do some checks?
  - Check if the crate-type is `["cdylib"]` during compilation (internal)
  - Check if the process is running with sufficient privileges (external)

**Comment by 7ap** [comment]
I think the adding the `init` attribute to `hax::main` would be fairly simple. I'll try to implement it and see how it goes.

**Comment by 7ap** [comment]
While I'm implementing this I had the idea of also being able to pass a `process` parameter, basically within the proc-macro everything for setting up an external will be done for you.

I think this idea can be augmented to eliminate a *ton* of boilerplate, for both external and internal.

Think of this example (external):
```rs
#[hax::main(process = "csgo.exe")]
fn entrypoint() {
    let value = hax.read::<i32>(0x123456);
}
```

Expanded to:
```rs
fn main() {
    // "hax context" will be created; you can use hax.????? to interact with it

    // ...
}
```

I think during creation of the "hax context" it could do checks depending on the feature flag (internal: check if crate-type is cdylib, external: check if process is running as admin, etc), and setup everything you need.

Maybe make the "hax context" globally accessible?

What do you think @not-matthias?

**Comment by not-matthias** [comment]
I like that approach. But we have to carefully think about whether or not to make the `hax_ctx` (hax context) static or pass it as parameter. 


**Static**: 
- Could lead to problems if multiple threads access it (reading/writing at the same time)
- Could this lead to some problems in very specific applications? We have to be careful if `lazy_static` or `once_cell` uses TLS callbacks. Cause often people try to do weird stuff with the dll (e.g. split it up or make it as small as possible) to get around detections. 

**Parameter**:
- If we pass ownership to the entrypoint, then we don't have any of those issues. The developer can store it globally if they want to. 
- Downside to this is, that you have to pass it around. 

**Both**: We could try and implement both, but I think this would lead to very confusing errors and other problems.

We should really ask more people what they think about those solutions and which they like best.

**Comment by 7ap** [comment]
Yep, that's my main concern with doing something like that. Definitely doing polls would be the best approach here.

Although, we could just have it default to a parameter unless the developer adds a `ctx = "global"` or something in the `hax::main` attribute. Something like `#[hax::main(process = "csgo.exe", ctx = "global")]`?

Another issue is that this already proving to be pretty difficult for me to add `init = ...`. I need to do a lot more research on proc-macros, however there is pretty limited documentation about them.

**Comment by not-matthias** [comment]
My primary concern against it, is having two different implementations which generally makes it confusing for the user. 

If one looks at an example which uses a global context, I imagine it can be confusing when you don't know about parameter contexts (especially if we only enable one by default)

**Comment by 7ap** [comment]
> My primary concern against it, is having two different implementations which generally makes it confusing for the user.

I didn't take that into account, you're right, it definitely would be confusing to the user. I'm trying to mentally figure out which one would be better for the user and it's confusing me xD

**Comment by not-matthias** [comment]
Allow to pass your own memory:

```
fn setup_memory -> impl MemoryRead + MemoryWrite {
    // ...
}

#[hax_macros::main(process = "ac_client")]
#[hax_macros::memory(setup_memory)]
fn main(memory: Memory) {
    println!("Hello, world!");
}
```
